src/effects/VST3/VST3Wrapper.cpp

```diff
@@ -0,0 +1,653 @@
+}
+
+void VST3Wrapper::EndParameterEdit()
+{
+   static_cast<ComponentHandler*>(mComponentHandler.get())->SetAccess(nullptr);
+}
+
+
+
+Steinberg::int32 VST3Wrapper::GetLatencySamples() const
+{
+   return mAudioProcessor->getLatencySamples();
+}
+
+EffectSettings VST3Wrapper::MakeSettings()
+{
+   return EffectSettings::Make<VST3EffectSettings>();
+}
+
+void VST3Wrapper::LoadSettings(const CommandParameters& parms, EffectSettings& settings)
+{
+   auto& vst3settings = GetSettings(settings);
+   vst3settings.processorState.reset();
+   vst3settings.controllerState.reset();
+   if(parms.HasEntry(processorStateKey))
+   {
+      vst3settings.processorState = parms.Read(processorStateKey);
+      if(parms.HasEntry(controllerStateKey))
+         vst3settings.controllerState = parms.Read(controllerStateKey);
+   }
+}
+
+void VST3Wrapper::SaveSettings(const EffectSettings& settings, CommandParameters& parms)
+{
+   const auto& vst3settings = GetSettings(settings);
+
+   if(vst3settings.processorState.has_value())
+      parms.Write(processorStateKey, *vst3settings.processorState);
+   if(vst3settings.controllerState.has_value())
+      parms.Write(controllerStateKey, *vst3settings.controllerState);
+}
+
+void VST3Wrapper::LoadUserPreset(const EffectDefinitionInterface& effect, const RegistryPath& name, EffectSettings& settings)
+{
+   auto& vst3settings = GetSettings(settings);
+
+   vst3settings.processorState.reset();
+   vst3settings.controllerState.reset();
+
+   wxString processorStateStr;
+   if(GetConfig(effect, PluginSettings::Private, name, processorStateKey, processorStateStr, wxEmptyString))
+   {
+      vst3settings.processorState = processorStateStr;
+      wxString controllerStateStr;
+      if(GetConfig(effect, PluginSettings::Private, name, controllerStateKey, controllerStateStr, wxEmptyString))
+         vst3settings.controllerState = controllerStateStr;
+   }
+}
+
+void VST3Wrapper::SaveUserPreset(const EffectDefinitionInterface& effect, const RegistryPath& name, const EffectSettings& settings)
+{
+   using namespace Steinberg;
+
+   const auto& vst3settings = GetSettings(settings);
+   if(vst3settings.processorState.has_value())
+   {
+      SetConfig(effect, PluginSettings::Private, name, processorStateKey, *vst3settings.processorState);
+      if(vst3settings.controllerState.has_value())
+         SetConfig(effect, PluginSettings::Private, name, controllerStateKey, *vst3settings.controllerState);
+   }
+}
+
+void VST3Wrapper::AssignSettings(EffectSettings& dst, EffectSettings&& src)
+{
+   if(!dst.has_value() )
+      dst = src;
+}
+
+void VST3Wrapper::CopySettingsContents(const EffectSettings& src, EffectSettings& dst, SettingsCopyDirection copyDirection)
+{
+   if(copyDirection == SettingsCopyDirection::MainToWorker)
+   {
+      //Don't allocate in worker
+      std::swap(
+         GetSettings(*const_cast<EffectSettings*>(&src)).parameterChanges,
+         GetSettings(dst).parameterChanges
+      );
```

Comment:
You have simply deleted the clearing of the `std::map` instead of the more complicated suggstion I made to use a map of unchanging size with `std::optional` values in it.
```suggestion
      );
```
```suggestion
      );
```
```suggestion
      );
```
```suggestion
      );
```

There is also a swapping of map contents.

I wonder whether this will reall work right.  It would be that the old contents swapped away might later be incorrectly swapped in again, if `CopySettingsContents` is called again before the main thread sends an updated message.

So I think the right thing to do, which might fix some bugs with real time adjustment of dials, is as I suggested.  More explicitly:

- `MakeSettings` makes a map from each parameter id to `nullopt`
- `performEdit` emplaces values
- `CopySettingsContents` copies each value and assigns `nullopt` to the source (note, std::move from the optional will NOT make the source null opt)

If you are experiencing some bugs with adjustment of sliders, then try this instead.

Reply:
The more complicated suggestion will not work because:
a) `MakeSettings` can't make such map because to do so it would need to instantiate `IEditController`
b) Assigning `nullopt` or clearing the map turned out to be pointless, considering that the real purpose is to clear setting on the main thread side (so that subsequent `Get` see it empty) and not the intermediate objects, which could be achieved only by using `Flush`

For now I guess it's not a big problem that main thread remebers latest values of all parameters that have changed

> It would be that the old contents swapped away might later be incorrectly swapped in again, if CopySettingsContents is called again before the main thread sends an updated message.

I thought that counter prevents that, isn't it?